include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -396,36 +396,114 @@ make_zip_iterator(std::tuple<_Tp...> __arg)
     return zip_iterator<_Tp...>(__arg);
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// struct UnaryFuncWrapperImplBaseTrviallyConstructible - predicate storage
+// if it's trivially constructible.
+template <typename _UnaryFunc>
+struct UnaryFuncWrapperImplBaseTrviallyConstructible
+{
+    _UnaryFunc pred;
+
+    UnaryFuncWrapperImplBaseTrviallyConstructible() = default;
+    UnaryFuncWrapperImplBaseTrviallyConstructible(_UnaryFunc _unary_func) : pred(_unary_func) {}
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// struct UnaryFuncWrapperImplBaseNonrviallyConstructible - predicate storage
+// if it's not trivially constructible.
+template <typename _UnaryFunc>
+struct UnaryFuncWrapperImplBaseNonrviallyConstructible
+{
+    _UnaryFunc pred;
+
+    UnaryFuncWrapperImplBaseNonrviallyConstructible() : pred{}
+    {
+    }
+
+    UnaryFuncWrapperImplBaseNonrviallyConstructible(_UnaryFunc _unary_func) : pred(_unary_func)
+    {
+    }
+};
+
+template <typename _UnaryFunc>
+using UnaryFuncWrapperImplBase = ::std::conditional_t<
+    ::std::is_trivially_constructible_v<_UnaryFunc>,
+    UnaryFuncWrapperImplBaseTrviallyConstructible<_UnaryFunc>,
+    UnaryFuncWrapperImplBaseNonrviallyConstructible<_UnaryFunc>>;
+
+////////////////////////////////////////////////////////////////////////////////
+// struct UnaryFuncWrapperImplCopyable - predicat storage
+// if it's trivially copyable
+template <typename _UnaryFunc>
+struct UnaryFuncWrapperImplCopyable : UnaryFuncWrapperImplBase<_UnaryFunc>
+{
+    UnaryFuncWrapperImplCopyable() = default;
+    UnaryFuncWrapperImplCopyable(_UnaryFunc __unary_func) : UnaryFuncWrapperImplBase<_UnaryFunc>(__unary_func) {}
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// struct UnaryFuncWrapperImplNotCopyable - predicat storage
+// if it's not trivially copyable
+template <typename _UnaryFunc>
+struct UnaryFuncWrapperImplNotCopyable : UnaryFuncWrapperImplBase<_UnaryFunc>
+{
+    UnaryFuncWrapperImplNotCopyable() = default;
+    UnaryFuncWrapperImplNotCopyable(_UnaryFunc __unary_func) : UnaryFuncWrapperImplBase<_UnaryFunc>(__unary_func) {}
+
+    UnaryFuncWrapperImplNotCopyable(const UnaryFuncWrapperImplNotCopyable<_UnaryFunc>&)
+    {
+        // do nothing
+    }
+    UnaryFuncWrapperImplNotCopyable(UnaryFuncWrapperImplNotCopyable<_UnaryFunc>&&)
+    {
+        // do nothing
+    }
+
+    UnaryFuncWrapperImplNotCopyable&
+    operator=(const UnaryFuncWrapperImplNotCopyable<_UnaryFunc>&)
+    {
+        // do nothing
+        return *this;
+    }
+
+    UnaryFuncWrapperImplNotCopyable&
+    operator=(UnaryFuncWrapperImplNotCopyable<_UnaryFunc>&&)
+    {
+        // do nothing
+        return *this;
+    }
+};
+
+template <typename _UnaryFunc>
+using UnaryFuncWrapper =
+    ::std::conditional_t<::std::is_trivially_copyable_v<_UnaryFunc>, UnaryFuncWrapperImplCopyable<_UnaryFunc>,
```

Comment:
`::std::is_trivially_copyable_v<_UnaryFunc>` should instead be 
`::std::is_trivially_copy_assignable_v<_UnaryFunc>`  or maybe just `::std::is_copy_assignable_v<_UnaryFunc>` (needs more thought).
However, `::std::is_trivially_copyable_v` does not imply this explicitly, as discussed offline.